releasenotes/notes/apm-sql-cache-03c32a5e90b21674.yaml

```diff
@@ -0,0 +1,13 @@
+# Each section from every releasenote are combined when the
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    APM: An SQL query obfuscation cache was added under the feature flag
+    DD_APM_FEATURES=sql_cache. In most cases where SQL queries are repeated
```

Comment:
Should this be enabled by default? What's the argument towards making it one more thing that the customer has to configure manually?

Reply:
The problem is that if the queries are not repeated, this starts working against us and instead slows things down. As an example, if the queries are never repeated it adds another 20-30% extra time per query obfuscation. We want to start slowly and then as we test this more and feel more comfortable, start enabling by default and even consider a system where the cache disables itself after a lot of misses.

Reply:
Here's an extreme result of how query obfuscation looks like with the cache implemented, but with no hits:
```
$ benchstat -delta-test none old.txt new.txt
name                         old time/op    new time/op    delta
ObfuscateSQLString/random-8    1.72µs ± 0%    2.63µs ± 0%  +52.99%
name                         old alloc/op   new alloc/op   delta
ObfuscateSQLString/random-8      720B ± 0%     1063B ± 0%  +47.64%
name                         old allocs/op  new allocs/op  delta
ObfuscateSQLString/random-8      22.0 ± 0%      26.0 ± 0%  +18.18%
```
One idea is to calculate a tipping point for the cache and automatically disable it when it is hit, but the problem is that the tipping point varies based on query length and number of tokens so the system would have to be smarted and average out query lengths which introduces even more complexity.

I hope this helps.